Determine install/upgrade version script

jhadvig opened the pull request:
This script will determine the GA and latest packages for previous version of package that will be on input.
EG:
input: ` origin-1.4.1-1.el7`
output:
 - ORIGIN_INSTALL_VERSION file that includes previous GA version-> `1.3.0-3.el7.x86_64`
 - ORIGIN_UPGRADE_VERSION file that includes previous latest version-> `1.3.3-1.el7.x86_64`

The script will return pre-release versions of packages if there are no released ones or fail.

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,55 @@
+input_pkg_version_list = map(int, input_pkg_obj.version.split("."))
+search_version_list = input_pkg_version_list[:]
+search_version_list[1] = search_version_list[1] -1
+search_version = ".".join(map(str, search_version_list)[:2])
```
Maybe the `rpm` package has utilities for this?

jhadvig replied:
Yeah I was checking particularly for that but haven't found anything. But will double check :+1:

stevekuznetsov replied:
If not, I think this could be written in such a way to be a little more clear:

```py
major, minor, *rest = input_pkg.version.split('.')
previous_version = '.'.join(major, minor+1, *rest)
```

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,55 @@
+search_version = ".".join(map(str, search_version_list)[:2])
+
+yb = yum.YumBase()
+yb.conf.disable_excludes = ["all"]
```
Add a comment why this is necessary

jhadvig replied:
Yeah will add bunch of comments

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,55 @@
+
+def determine_install_upgrade_versions(version_list):
+	write_package_version(version_list[0], "install")
+	if (len(version_list) > 1) and (version_list[0] is not version_list[len(version_list)-1]):
```
Why do we need the second check?

stevekuznetsov replied:
Also, do not use `is not` -- it will check to see if those two items are literally in the same memory address, not if they are equal.

jhadvig replied:
well I thought that it might happen that there will be pkg duplicity when two repos for some reason would contain the same version of pkg. But not 100% sur eif that could occur...

stevekuznetsov replied:
Let's worry about it if it happens -- for now I don't think we need to.

jhadvig commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,55 @@
+import yum
+import sys
+import os.path
+from version_utils import rpm
```
They are... at least I was using them and needed to install the `version_utils` via pip

stevekuznetsov replied:
Hmm... I think we need to be using what's installed with

```
yum install rpm-python
```

jhadvig commented:
@stevekuznetsov uptedate.. tested it with most of the versions of `origin` and `openshift-ansible`
PTAL

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,56 @@
+    print "[ERROR] No NEVRA provided!"
+    sys.exit(3)
+input_pkg = sys.argv[1]
+input_pkg_list = rpmutils.splitFilename(input_pkg)
```
break this out

```py
name, version, release, epoch, arch =  rpmutils.splitFilename(input_pkg)
```

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,56 @@
+input_pkg_version_list = map(int, input_pkg_list[1].split("."))
+search_version_list = input_pkg_version_list[:]
+search_version_list[1] = search_version_list[1] -1
+search_version = ".".join(map(str, search_version_list)[:2])
```
still think the readability approach is better 

```py
major, minor, *rest = version.split('.')
search_version = '.'.join(major, minor - 1, *rest)
```

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,56 @@
+# Write install and upgrade versions of desired package into a file called "package_name"_INSTALL_UPGRADE_VERSION
+# The upgrade version will be written into the file only in case there is more then one version of packages previous
+# minor release. 
+def determine_install_upgrade_versions(version_list):
```
```py
def write_version_vars(path, install_version, upgrade_version=None):
```

stevekuznetsov commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,55 @@
+pre_release_pkgs = []
+
+available_pkgs = yb.doPackageLists('available', patterns=[pkg_name], showdups=True).available
+available_pkgs.sort(key=lambda x: x.version)
```
Shouldn't this use the `rpm.versionCompare()` ?

jhadvig commented:
@stevekuznetsov updated

stevekuznetsov commented:
Only thing we're missing is a suite of test cases

jhadvig commented:
@stevekuznetsov TCs added

stevekuznetsov commented on hack/determine_install_upgrade_version_test.py:
```diff
@@ -0,0 +1,41 @@
+	MATCHING_1_3 = ["1.3.1-1.el7.x86_64", "1.3.2-3.el7.x86_64"]
+	MATCHING_1_2 = ["1.2.0-0.el7.x86_64", "1.2.5-0.el7.x86_64"]
+
+	def test_get_matching_versions(self):
```
Prefer to hard-code test data in tests rather than in more global scope.

stevekuznetsov commented on hack/determine_install_upgrade_version_test.py:
```diff
@@ -0,0 +1,41 @@
+	MATCHING_1_2 = ["1.2.0-0.el7.x86_64", "1.2.5-0.el7.x86_64"]
+
+	def test_get_matching_versions(self):
+		self.assertEqual(get_matching_versions(self.PKG_NAME, test_pkgs_objs, "1.5"), self.MATCHING_1_5)
```
Let's break these up into specific tests and name them by the intent that they surface. For instance, this test could be:

```py
def test_get_matching_versions_only_prelease(self):
    """ when only pre-release versions exist, they are returned """
    self.assertEqual(
        get_matching_versions('origin', ['1.5.0-0.4.el7.x86_64'], '1.5'),
        ['1.5.0-0.4.el7.x86_64']
    )
```

Then you could pair it with:
```py
def test_get_matching_versions_release_and_prelease(self):
    """ when both release and pre-release versions exist, only release versions are returned """
    self.assertEqual(
        get_matching_versions('origin', ['1.5.0-0.4.el7.x86_64','1.5.0-1.4.el7.x86_64'], '1.5'),
        ['1.5.0-1.4.el7.x86_64']
    )
```

And follow up with the other tests you have written -- when more than one version exists, all are returned, etc.

stevekuznetsov replied:
I guess instead of the strings you'd need literal `TestPackage` objects or `dict`s

stevekuznetsov commented on hack/determine_install_upgrade_version_test.py:
```diff
@@ -0,0 +1,41 @@
+
+
+if __name__ == '__main__':
+	test_pkgs = ["origin-1.2.0-0.el7.x86_64",
```
Always prefer to place test data as a literal object in the test scope. Otherwise, you start giving tests more than one reason to fail, and that makes it harder to debug failures.

kargakis commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,65 @@
+def write_version_vars(version_list):
+	file_name = pkg_name.upper() + "_INSTALL_UPGRADE_VERSION"
+	if os.path.isfile(file_name):
+		os.remove(file_name)
```
Wondering if this works since you only import `os.path`...

jhadvig commented:
@stevekuznetsov @kargakis 
did:
- refactored the TC's  
- added updated `test_branch_origin_extended_conformance_install_update.yml` that will work with new `determine_install_upgrade_version.py`
- removed old `get_available_pkgs.py`

PTAL

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,100 @@ extensions:
+        pkg_name=$( cat ./PKG_NAME )
+        if [ $pkg_name = "atomic-openshift" ]
+        then
+          sed -i "s/deployment_type:.*/deployment_type: \"openshift-enterprise\" /g" inventory/group_vars/OSEv3/general.yml
```
pass as `-e` to `ansible-playbook`

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,100 @@ extensions:
+        ORIGIN_INSTALL_MINOR_VERSION=$( echo ${ORIGIN_INSTALL_VERSION} | awk -F'.' '{print $2}' )
+        echo "=== Installing ${pkg_name}-${ORIGIN_INSTALL_VERSION} ==="
+        ansible-playbook -vv --become --become-user root --connection local -i inventory/ /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-node/network_manager.yml
+        ansible-playbook -vv --become --become-user root --connection local -i inventory/ /usr/share/ansible/openshift-ansible/playbooks/byo/config.yml -e openshift_pkg_version="-${ORIGIN_INSTALL_VERSION}" -e enable_excluders=false
```
Why no excluders?

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,100 @@ extensions:
+          echo "=== Updating openshift-ansible to openshift-ansible-${ATOMIC_OPENSHIFT_UTILS_UPGRADE_VERSION} ==="
+          sudo yum upgrade -y atomic-openshift-utils-${ATOMIC_OPENSHIFT_UTILS_UPGRADE_VERSION}
+          echo "=== Updating ${pkg_name} to ${ORIGIN_UPGRADE_VERSION} ==="
+          ansible-playbook -vv --become --become-user root --connection local --inventory inventory/ /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/upgrades/v3_${ORIGIN_INSTALL_MINOR_VERSION}/upgrade.yml -e openshift_pkg_version="-${upgrade_version}" -e oreg_url='openshift/origin-${component}:'"$( echo ${ORIGIN_UPGRADE_VERSION} | awk -F'-' '{print $1}' )"
```
Again if you're putting `NVR`together in the var or whatever, also provide the specific fields that we need -- the Bash should be entirely minimal.

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,100 @@ extensions:
+        then
+          upgrade_pkgs="$( cat ./OPENSHIFT_ANSIBLE_PKGS )"
+        fi
+        sudo yum upgrade -y ${upgrade_pkgs}
```
Why aren't these version-pinned to what we built?

jhadvig replied:
forgot that we need to test specific verions

stevekuznetsov commented on hack/determine_install_upgrade_version_test.py:
```diff
@@ -0,0 +1,95 @@
+	"Test for `determine_install_upgrade_version.py`"
+
+	def setUp(self):
+		test_pkgs = ["origin-1.2.0-0.el7.x86_64",
```
Prefer not to use global state, list the specific things you need in each test as necessary

jhadvig replied:
but we need to test that right versions of install/upgrade pkgs are picked, from a list of diferent versions and not just releases. For that reason I've created a list of pkg versions that simulates the output that would be provided by yum. For picking the right instlal/upgrade version from the matching versions I've created the `GetInstallUpgradeVersionTestCase`

stevekuznetsov commented on hack/determine_install_upgrade_version_test.py:
```diff
@@ -0,0 +1,95 @@
+	"Test for `determine_install_upgrade_version.py`"
+
+	def test_get_install_upgrade_with_multiple_matching_release_versions(self):
+		""" when install and upgrade version are picked from multiple matching versions of released versions are """
```
did you forget the end of this comment?

stevekuznetsov commented on hack/determine_install_upgrade_version_test.py:
```diff
@@ -0,0 +1,95 @@
+		self.assertEqual(get_install_upgrade_version(matching_versions), install_upgrade_versions)
+
+	def test_get_install_upgrade_with_single_matching_release_version(self):
+		""" when install and upgrade version are picked from single matching versions of released versions are """
```
copy pasta

jhadvig replied:
no, the comments should differ from the rest of them.. check them

stevekuznetsov replied:
hm -- very hard to see the difference. I'd suggest one of two things:

 - your method names / test names are verbose and expressive. if you think you captured the entirety of the test reason in the name, get rid of the comment
 - emphasize the differences... so this might read "when only a single matching version is present..."

jhadvig commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,100 @@ extensions:
         name = OpenShift Ansible Release from Local Source
         EOR
         sudo cp ./openshift-ansible-local-release.repo /etc/yum.repos.d
+        basename $(ls ${tito_tmp_dir}/noarch/ | grep atomic-openshift-utils ) .rpm >> /data/src/github.com/openshift/aos-cd-jobs/OPENSHIFT_ANSIBLE_BUILT_VERSION
```
`basename "${tito_tmp_dir}"/noarch/atomic-openshift-utils*.rpm .rpm >> /data/src/github.com/openshift/aos--cd-jobs/OPENSHIFT_ANSIBLE_BUILT_VERSION`

have to add the adidtional `.rpm` argument canuse I need only the name without the extension

stevekuznetsov replied:
ack

kargakis commented on hack/determine_install_upgrade_version.py:
```diff
@@ -0,0 +1,74 @@
+
+if __name__ == "__main__":
+	if len(sys.argv) != 2:
+	    print "[ERROR] No NEVRA provided!"
```
print errors to stderr

kargakis commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -6,63 +6,19 @@ extensions:
-        sed -i 's|go/src|data/src|' _output/local/releases/rpms/origin-local-release.repo
-        sudo cp _output/local/releases/rpms/origin-local-release.repo /etc/yum.repos.d/
+        source hack/lib/build/rpm.sh; source hack/lib/log/output.sh; source hack/common.sh; source hack/lib/build/rpm.sh; source hack/lib/util/text.sh
+        git log -1 --pretty=%h >> /data/src/github.com/openshift/aos-cd-jobs/ORIGIN_COMMIT
```
Why is this appending instead of replacing?

jhadvig commented:
@stevekuznetsov check & merge :grimacing:

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,102 @@ extensions:
+        source ATOMIC-OPENSHIFT-UTILS_INSTALL_UPGRADE_VERSION
+        echo "=== Installing ${pkg_name}-${ORIGIN_INSTALL_VERSION} ==="
+        ansible-playbook -vv --become --become-user root --connection local -i inventory/ /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-node/network_manager.yml -e deployment_type=$( cat ./DEPLOYMENT_TYPE)
+        ansible-playbook -vv --become --become-user root --connection local -i inventory/ /usr/share/ansible/openshift-ansible/playbooks/byo/config.yml -e openshift_pkg_version="-${ORIGIN_INSTALL_VERSION}" -e enable_excluders=false -e deployment_type=$( cat ./DEPLOYMENT_TYPE)
```
Why no excluders here?

jhadvig replied:
The excluders are fixed and now they are installed and enabled upon the install process. During the upgrade process they are disabled so no need to disable them now

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,102 @@ extensions:
+          echo "=== Updating openshift-ansible to openshift-ansible-${ATOMIC_OPENSHIFT_UTILS_UPGRADE_VERSION} ==="
+          sudo yum upgrade -y atomic-openshift-utils-${ATOMIC_OPENSHIFT_UTILS_UPGRADE_VERSION}
+          echo "=== Updating ${pkg_name} to ${ORIGIN_UPGRADE_VERSION} ==="
+          ansible-playbook -vv --become --become-user root --connection local --inventory inventory/ /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/upgrades/v3_${ORIGIN_UPGRADE_MINOR_VERSION}/upgrade.yml -e openshift_pkg_version="-${ORIGIN_UPGRADE_VERSION}" -e deployment_type=$( cat ./DEPLOYMENT_TYPE) -e oreg_url='openshift/origin-${component}:'"${ORIGIN_UPGRADE_VERSION_PKG_ ...
```
Nit, but we should see if this works to make multiline shell -- would read better:

```sh
upgrade_playbook="/usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/upgrades/v3_${ORIGIN_UPGRADE_MINOR_VERSION}/upgrade.yml"
ansible-playbook -vv --become           \
                 --become-user root     \
                 --connection local     \
                 --inventory inventory/ \
                  "${upgrade_playbook}" \                   \
                 -e deployment_type=$( cat ./DEPLOYMENT_TYPE)          \
                 -e openshift_pkg_version="-${ORIGIN_UPGRADE_VERSION}" \
                 -e oreg_url='openshift/origin-${component}:'"${ORIGIN_UPGRADE_VERSION_PKG_VERSION}" 
```

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,102 @@ extensions:
+        source ATOMIC-OPENSHIFT-UTILS_INSTALL_UPGRADE_VERSION
+        if [[ -v ATOMIC_OPENSHIFT_UTILS_UPGRADE_MINOR_VERSION && ${ATOMIC_OPENSHIFT_UTILS_UPGRADE_MINOR_VERSION} == 4 ]]
+        then
+          upgrade_openshift_ansible_pkgs="$( cat ./OPENSHIFT_ANSIBLE_PKGS )"
```
`packages_to_upgrade`?

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,102 @@ extensions:
+        fi
+        for pkg in "${upgrade_openshift_ansible_pkgs[@]}"
+        do
+          upgrade_openshift_ansible_pkgs="${upgrade_openshift_ansible_pkgs} ${pkg}-${ATOMIC_OPENSHIFT_UTILS_UPGRADE_MINOR_VERSION}"
```
`versioned_packages_to_upgrade`?

jhadvig commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,102 @@ extensions:
+        echo $pkg_name > PKG_NAME
+        echo "openshift-ansible openshift-ansible-callback-plugins openshift-ansible-docs openshift-ansible-filter-plugins openshift-ansible-lookup-plugins openshift-ansible-playbooks openshift-ansible-roles" > OPENSHIFT_ANSIBLE_PKGS
+        sudo python hack/determine_install_upgrade_version.py $( cat OPENSHIFT_ANSIBLE_BUILT_VERSION )
+        source ATOMIC-OPENSHIFT-UTILS_INSTALL_UPGRADE_VERSION
```
So you suggest just to print out the versions from the script ?
Well Im not the biggest fan of both versions but I sticked to the idea to do less magic with bash and more with python scripts, thats why I'm writing it directly to file, so in other steps we just source the file and we are good.
But I can go with:
```
sudo hack/determine_install_upgrade_version.py "$( cat OPENSHIFT_ANSIBLE_BUILT_VERSION )" > OPENSHIFT_ANSIBLE_VARS
source OPENSHIFT_ANSIBLE_VARS
```

jhadvig replied:
so ?

stevekuznetsov replied:
Yeah, writing it to `stdout` makes sense -- it should look very similar in the Python script but make the flow of data a little more clear in the Bash

jhadvig commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +28,102 @@ extensions:
+        sudo python hack/determine_install_upgrade_version.py $( cat OPENSHIFT_ANSIBLE_BUILT_VERSION )
+        source ATOMIC-OPENSHIFT-UTILS_INSTALL_UPGRADE_VERSION
+        openshift_ansible_pkgs_versions=""
+        openshift_ansible_pkgs="atomic-openshift-utils"
```
yeah but wanted to be more explicit ... wanna rename it ?

jhadvig commented:
PR updated
Testing PR: https://ci.openshift.redhat.com/jenkins/view/All/job/test_branch_origin_extended_conformance_install_update/138/console

@stevekuznetsov PTAL

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +29,138 @@ extensions:
+                          -i inventory/      \
+                          /usr/share/ansible/openshift-ansible/playbooks/byo/config.yml \
+                          -e openshift_pkg_version="-${ORIGIN_INSTALL_VERSION}"         \
+                          -e enable_excluders=false                                     \
```
Based on our conversation this morning, we can drop this, right?

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +29,138 @@ extensions:
+        fi
+        for pkg in "${packages_to_install[@]}"
+        do
+          versioned_packages_to_install="${versioned_packages_to_install} ${pkg}-${ATOMIC_OPENSHIFT_UTILS_INSTALL_VERSION}"
```
nit, this can be 

```sh
versioned_packages_to_install+=" ${pkg}-${ATOMIC_OPENSHIFT_UTILS_INSTALL_VERSION}"
```

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +29,138 @@ extensions:
+          packages_to_upgrade="$( cat ./OPENSHIFT_ANSIBLE_PKGS )"
+        fi
+        versioned_packages_to_upgrade=""
+        for pkg in "${upgrade_openshift_ansible_pkgs[@]}"
```
Where is `upgrade_openshift_ansible_pkgs` from?

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +29,138 @@ extensions:
+        packages_to_upgrade="ansible atomic-openshift-utils"
+        source ORIGIN_VARS
+        source OPENSHIFT_ANSIBLE_VARS
+        if [[ -v ATOMIC_OPENSHIFT_UTILS_UPGRADE_MINOR_VERSION && ${ATOMIC_OPENSHIFT_UTILS_UPGRADE_MINOR_VERSION} == 4 ]]
```
Why do we only want to upgrade these if we're on minor .4 ?

jhadvig commented:
@stevekuznetsov 
I've added two aditional steps into the playbook:
1. that will print all the installed and updated versions of origin and openshift-ansible using `yum history`
2. that will check that proper images are on the cluster and that they are running

Regarding the `enable_excluders` variable that we spoke about... for now we have to keep them until 1.5-release pkg are re-built, backporting this change https://github.com/openshift/openshift-ansible/pull/3693
After that we can remove the variable. Sorry about the misunderstanding

kargakis commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -72,30 +29,141 @@ extensions:
+        pkg_name="origin"
+        git checkout sjb
+        echo $pkg_name > PKG_NAME
+        wget https://raw.githubusercontent.com/jhadvig/aos-cd-jobs/determin_version_test/hack/determine_install_upgrade_version.py -P hack
```
s/jhadvig/openshift/

Or this won't work until it merges?

jhadvig replied:
yeah.. it wont until its merged.. wanted to remove after LGTM

stevekuznetsov commented on config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -57,45 +135,44 @@ extensions:
+        sudo yum history info $( cat ./PKG_NAME )
     - type: "script"
-      title: "update origin"
+      title: "get history of the openshift-ansible package"
```
Why not as generated artifacts?

jhadvig commented:
:champagne:
